JSON encoder for Number values produces duplicate fields

## Description

The JSON encoder's `encodeScalar` method has a special case for encoding [`Number`]() values, but the special case doesn't return so the value is encoded a string as well.

This causes the same field to be encoded twice, introducing duplicate fields on the JSON object and causing errors when unmarshalling.

<details>
<summary>Relevant section of the <code>encodeScalar</code> method</summary>

https://github.com/aws/smithy-go/blob/16d913d93001992645c0f54e3ce9fa6be287b70a/document/json/encoder.go#L217-L230

</details>

## Impact

I have a use-case where I receive and unmarshal a document containing `Number` values, and then pass that document back to the API again. Unfortunately this issue causes that second API call to error because of the malformed JSON input!

## Reproduction

```go
package main

import (
	"testing"

	smithydocument "github.com/aws/smithy-go/document"
	smithydocumentjson "github.com/aws/smithy-go/document/json"
)

func TestEncodeNumber(t *testing.T) {
	res, err := smithydocumentjson.NewEncoder().Encode(map[string]interface{}{
		// Example of tool call parameters unmarshalled into map[string]interface{}
		// Smithy encodes this parameter twice - once as a JSON number and once as a JSON string.
		// https://github.com/aws/smithy-go/blob/16d913d93001992645c0f54e3ce9fa6be287b70a/document/json/encoder.go#L217-L230
		"x": smithydocument.Number("2"),
	})

	if err != nil {
		t.Fatalf("MarshalSmithyDocument returned error: %v", err)
	}
	expected := `{"x":2}`
	if string(res) != expected {
		t.Errorf("Expected JSON output to be %s, got %s", expected, string(res))
	}
}
```

Produces:

```
=== RUN   TestEncodeNumber
    smithy_test.go:23: Expected JSON output to be {"x":2}, got {"x":2,"x":"2"}
--- FAIL: TestEncodeNumber (0.00s)

FAIL
```

--- a/document/json/encoder.go
+++ b/document/json/encoder.go
@@ -232,6 +232,7 @@ func (e *Encoder) encodeScalar(vp valueProvider, rv reflect.Value) error {
 			return &document.InvalidMarshalError{Message: fmt.Sprintf("invalid number literal: %s", number)}
 		}
 		vp.GetValue().Write([]byte(number))
+		return nil
 	}
 
 	switch rv.Kind() {

--- a/document/json/encoder_test.go
+++ b/document/json/encoder_test.go
@@ -93,6 +93,44 @@ func TestDeterministicMapOrder(t *testing.T) {
 	}
 }
 
+func TestDocumentNumberEncodesAsNumber(t *testing.T) {
+	encoder := json.NewEncoder()
+
+	cases := map[string]struct {
+		input    interface{}
+		expected string
+	}{
+		"integer": {
+			input:    map[string]interface{}{"x": document.Number("42")},
+			expected: `{"x":42}`,
+		},
+		"float": {
+			input:    map[string]interface{}{"x": document.Number("42.0")},
+			expected: `{"x":42.0}`,
+		},
+		"negative zero": {
+			input:    map[string]interface{}{"x": document.Number("-0.0")},
+			expected: `{"x":-0.0}`,
+		},
+		"scientific notation": {
+			input:    map[string]interface{}{"x": document.Number("1.5e+10")},
+			expected: `{"x":1.5e+10}`,
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			b, err := encoder.Encode(tc.input)
+			if err != nil {
+				t.Fatalf("Encode() error: %v", err)
+			}
+			if string(b) != tc.expected {
+				t.Errorf("Encode() = %s, want %s", b, tc.expected)
+			}
+		})
+	}
+}
+
 func testEncode(t *testing.T, tt testCase) {
 	t.Helper()
 

--- a/encoding/json/value.go
+++ b/encoding/json/value.go
@@ -106,6 +106,11 @@ func (jv Value) BigInteger(v *big.Int) {
 
 // BigDecimal encodes v as JSON value
 func (jv Value) BigDecimal(v *big.Float) {
+	if v.Sign() == 0 && v.Signbit() {
+		// Preserve negative zero sign which Int64() would lose.
+		jv.w.Write([]byte("-0"))
+		return
+	}
 	if i, accuracy := v.Int64(); accuracy == big.Exact {
 		jv.Long(i)
 		return

--- a/encoding/json/value_test.go
+++ b/encoding/json/value_test.go
@@ -138,6 +138,26 @@ func TestValue(t *testing.T) {
 			},
 			expected: "9223372036854775807",
 		},
+		"bigFloat negative zero preserves sign": {
+			setter: func(value Value) {
+				v := new(big.Float).SetFloat64(math.Copysign(0, -1))
+				value.BigDecimal(v)
+			},
+			expected: "-0",
+		},
+		"bigFloat negative integer": {
+			setter: func(value Value) {
+				v := new(big.Float).SetFloat64(-7)
+				value.BigDecimal(v)
+			},
+			expected: "-7",
+		},
+		"bigFloat positive zero": {
+			setter: func(value Value) {
+				value.BigDecimal(new(big.Float))
+			},
+			expected: "0",
+		},
 	}
 	scratch := make([]byte, 64)
 

